### ISSUE TITLE
Cypress code coverage report

### ISSUE THREAD

**Issue by pzvalueadd** [nrwl/nx#2438]
Is it possible to generate Cypress code coverage report? I'm trying to set it up, but so far without success.

**Comment by Flaxline** [comment]
Did you make any progress on this?
I'm also trying to setup code coverage for a nx-repo with Angular, where it seems my 2 biggest problems are Angular/TS and the structure of the repository.
My goal would be to have the coverage for e2e-Tests against a Storybook, as they're now supported and generated by nx.

I've tried a lot of different setups with manual instrumentation of the code, and my, I think, best result so far was:

- e2e-Tests against the application
- the application manually built (not a prod-build), instrumented with nyc and started separately to the tests
- I had instrumented code and ran tests against it, but the report generation task (by cypress coverage plugin) failed. Could be due to missing source-mapping

As also noted before, my biggest problems seem to be 

- I'm using Typescript and I need the code to be instrumented after transpilation, but before execution. 
- I'm using the NX-Repo structure, EVERY sample, working or not, uses a global /cypress-Folder. So I assume https://github.com/cypress-io/code-coverage/issues/18 is also a huge problem there.

I plan to try to set up a global /cypress-Folder and configuring the code-coverage there, to see if at least my second assumption is correct. But I find it very hard to get resources on this topic, and on top most of them relating to react.

TL;DR:
Also trying, not being overly successful. Would appreciate some help, too.

**Comment by pzvalueadd** [comment]
Unfortunately I didn't make any progress so far :(

**Comment by Flaxline** [comment]
I finally figured out what I was doing wrong applying the steps from https://github.com/skylock/cypress-angular-coverage-example
Basically, if you are also using angular, you can follow that, but have to be very careful about the paths, especially in the webpack config.

I also plan on publishing a working example with NX-structure (and the storybook-based tests!) later (after I've finished up on the docs tbh).

**Comment by FERNman** [comment]
As I've very recently configured our project to generate cypress code-coverage reports and the example from @Flaxline didn't work for me, I thought I'd share the process here.

Basically, https://github.com/skylock/cypress-angular-coverage-example did the job. 

This will generate the coverage report in `apps/*-e2e/coverage`. If you want the report to be generated in the global coverage folder in the root of your repository instead (following the same convention as with unit tests), you have to add a `package.json` (at least until cypress-io/code-coverage#93 is implemented and additional config files are supported) containing at least the following:

```json
{
  ...
  "nyc": {
    "report-dir": "../../coverage/apps/<app_name>-e2e"
  },
  ...
}
```

This will generate the coverage in `/coverage/app-e2e/<app_name>-e2e`.

**Comment by masonembry** [comment]
Does anyone have any insight into how to setup coverage in an `nx` React app? I'd very much appreciate any guidance someone may have.

**Comment by maierson** [comment]
For React app frontend coverage:

This is as of nrlw 12.0.7 with nx cli installed locally \
Given an app `my-app` create the `my-app-e2e` cypress app.

---

`package.json`:

```json
"scripts": {
  ...
  // runs cypress for my-app-e2e
  "my-app:cy": "cross-env NODE_ENV=e2e nx e2e my-app-e2e",
  ...
},
...
"nyc": {
  "reporter": [
    "html",
    "json"
  ]
}
...
```

- Run:

```
  // optional if you want to only instrument code in the e2e environment
  // see `my-app:cy` script in package.json
  yarn add cross-env -D
```

---

`my-app-e2e`

- In `plugins/index.js` add:
  ```js
  module.exports = (on, config) => {
    ...
    // add this line to execute code coverage
    require("@cypress/code-coverage/task")(on, config);
    ...
    return config;
  };
  ```
- In `support/index.ts` add:
  ```ts
  import "@cypress/code-coverage/support";
  ...
  ```

---

`my-app`

Instrument for coverage:

- Run

```
yarn add babel-plugin-istanbul -D
```

- In `.babelrc` add:

```json
...
// this instruments your code for coverage before the tests run
"env": {
  "e2e": {
    "plugins": ["istanbul"],
  }
}
...
```

---

`my-lib-for-my-app`

Instrument for coverage every lib that `my-app` depends on and you want to get code coverage for:

- In `.babelrc` add:

```json
...
"env": {
  "e2e": {
    "plugins": ["istanbul"],
  }
}
...
```

Run `my-app-e2e` tests

  ```
  yarn my-app:cy
  ```

The `coverage` folder is created inside `my-app-e2e` folder.

**Comment by kerry-afs** [comment]
I know this is closed, but hoping someone who comes across this or has answered previously can provide some insight. 

Using @Flaxline 's solution unfortunately does not work for me at this point with recent versions of Webpack(5), Nx (14.x), Angular (14.x) and Cypress (10.x). 

Using `ngx-build-plus` and adding the `extraWebpackConfig` option for serving the app results in a schema mismatch as that property doesn't appear to be valid any longer. Additionally, `istanbul-instrumenter-loader` has a dependency on Webpack 4, so projects with Webpack 5 cannot use it.

I've toyed around with using `@jsdevtools/coverage-istanbul-loader` but ultimately, it doesn't appear that my Angular app is getting instrumented as the output for Cypress' code coverage is empty.

It's possible there's something wonky in my configuration below, but it isn't immediately clear to me. If anyone has any insight or a working code coverage configuration, I'd greatly appreciate it!

---
`package.json`
```
{
     ...
    "nyc": {
      "extends": "@istanbuljs/nyc-config-typescript",
      "all": true
    }
}
```
---
`coverage.webpack.json`
```
import * as path from 'path';

export default {
  module: {
    rules: [
      {
        test: /\.(js|ts)$/,
        loader: '@jsdevtools/coverage-istanbul-loader',
        options: { esModules: true },
        enforce: 'post',
        include: path.join(__dirname, 'apps', 'my-app', 'src'),
        exclude: [/\.(e2e|spec)\.ts$/, /node_modules/, /(ngfactory|ngstyle)\.js/],
      },
    ],
  },
};

```
---
`my-app > project.json`
```
{
  "build": {
    "executor": "@angular-builders/custom-webpack:browser",
    ...
    "configurations": {
      ...
      "development": {
        "buildOptimizer": false,
        "optimization": false,
        "vendorChunk": true,
        "extractLicenses": false,
        "sourceMap": true,
        "namedChunks": true,
        "customWebpackConfig": {
          "path": "tools/executors/cypress-coverage-report/coverage.webpack.ts"
        }
      }
    },
    ...  
  },
  "serve": {
    "executor": "@angular-devkit/build-angular:dev-server",
    "configurations": {
      "development": {
        "browserTarget": "my-app:build:development",
        "disableHostCheck": true
      }
    },
    "defaultConfiguration": "development"
  },
}
```
---

**Comment by DerHerrGammler** [comment]
i run into the same problem as kerry-afs and if someone of u got a solution for that would be very helpful

**Comment by danielzenzen** [comment]
is there a working solution for this?
